src/lib/storage/disk_size.rb

```diff
@@ -0,0 +1,260 @@
+        end
+
+        def ZiB(size)
+          DiskSize.new(size * (1024**6))
+        end
+
+        def YiB(size)
+          DiskSize.new(size * (1024**7))
+        end
+
+        def unlimited
+          DiskSize.new(-1)
+        end
+
+        def zero
+          DiskSize.new(0)
+        end
+
+        # Create a DiskSize from a parsed string.
+        # Valid formats:
+        #   42 GiB
+        #   42.00  GiB
+        #   42          (=> 42 kiB)
+        #   unlimited   (=> -1 == unlimited)
+        #
+        # Invalid:
+        #   42 GB    (supporting binary units only)
+        #
+        def parse(str)
+          str.strip!
+          return DiskSize.unlimited if str == UNLIMITED
+          size_str, unit = str.split(/\s+/)
+          raise ArgumentError, "Bad number: #{size_str}" if size_str !~ /^\d+\.?\d*$/
+          size = size_str.to_f
+          return DiskSize.new(size) if unit.nil?
+          DiskSize.new(size * unit_multiplier(unit))
+        end
+
+        alias_method :from_s, :parse
+        alias_method :from_human_readable, :parse
+
+        # Return the unit exponent for any of the known binary units ("kiB",
+        # "MiB", ...). The base of this exponent is 1024. The base unit is kiB.
+        #
+        def unit_exponent(unit)
+          index = UNITS.index(unit)
+          raise ArgumentError, "expected one of #{UNITS}" if index.nil?
+          index
+        end
+
+        # Return the unit multiplier for any of the known binary units ("kiB",
+        # "MiB", ...). The base unit is kiB.
+        #
+        def unit_multiplier(unit)
+          1024 ** unit_exponent(unit)
+        end
+      end
+
+      #
+      # Operators
+      #
+
+      def +(other)
+        if other.is_a?(Numeric)
+          DiskSize.new(@size_k + other)
+        elsif other.respond_to?(:size_k)
+          DiskSize.new(@size_k + other.size_k)
+        else
+          raise TypeError, "Numeric value or DiskSize expected"
+        end
+      end
+
+      def -(other)
+        if other.is_a?(Numeric)
+          DiskSize.new(@size_k - other)
+        elsif other.respond_to?(:size_k)
+          DiskSize.new(@size_k - other.size_k)
+        else
+          raise TypeError, "Numeric value or DiskSize expected"
+        end
+      end
+
+      def *(other)
+        if other.is_a?(Numeric)
+          DiskSize.new(@size_k * other)
+        else
+          raise TypeError, "Numeric value expected"
+        end
+      end
+
+      def /(other)
+        if other.is_a?(Numeric)
+          DiskSize.new(@size_k.to_f / other)
+        else
+          raise TypeError, "Numeric value expected"
+        end
+      end
+
+      #
+      # Other methods
+      #
+
+      def unlimited?
+        size_k == -1
+      end
+
+      def zero?
+        size_k == 0
+      end
+
+      # The Comparable mixin will get us operators < > <= >= == != with this
+      def <=>(other)
+        if other.respond_to?(:size_k)
+          return @size_k <=> other.size_k
```

Comment:
I think there is one logical error. Using -1 for unlimited mean, that unlimited is smaller then zero, which is from mathematical point of view invalid.